d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -867,29 +873,26 @@ public void run()
       for (ResourceType type : _resourceSubscribers.keySet())
       {
         Set<String> resources = new HashSet<>(getResourceSubscriberMap(type).keySet());
-        if (resources.isEmpty() && getWildcardResourceSubscriber(type) == null)
+        if (resources.isEmpty())
         {
           continue;
         }
-        ResourceType rewrittenType;
+
+        ResourceType rewrittenType = type;
         if (_subscribeToUriGlobCollection && type == ResourceType.D2_URI_MAP)
         {
           resources = resources.stream()
               .map(GlobCollectionUtils::globCollectionUrlForClusterResource)
               .collect(Collectors.toCollection(HashSet::new));
           rewrittenType = ResourceType.D2_URI;
         }
-        else
-        {
-          rewrittenType = type;
-        }
-        // If there is a wildcard subscriber, we should always send a wildcard request to the server.
-        if (getWildcardResourceSubscriber(type) != null)
-        {
-          resources.add("*");
-        }
         _adsStream.sendDiscoveryRequest(rewrittenType, resources);
       }
+
+      for (ResourceType type: _wildcardSubscribers.keySet())
+      {
+        _adsStream.sendDiscoveryRequest(type, Collections.singletonList("*"));
```

Comment:
The key in the map is always `D2_URI_MAP` here even if glob collections are enabled, so this will actually break because it will resubscribe to the map instead of the glob collection. You have to check `_subscribeToUriGlobCollection && type == ResourceType.D2_URI_MAP` and do the translation accordingly.

Reply:
good catch. Will do!

Reply:
fixed.